compute_endpoint/globus_compute_endpoint/endpoint/rabbit_mq/result_publisher.py

```diff
@@ -0,0 +1,358 @@
+        :param channel_close_window_limit: Limit of channel close events (within
+            ``channel_close_window_s``) before shutting down the thread.
+        """
+        self.queue_info = queue_info
+
+        # how often to check for work; every `poll_period_s`, the `_event_watcher`
+        # method will handle any outstanding work.
+        self.poll_period_s = max(0.01, poll_period_s)
+
+        # how many times to attempt connection before giving up and shutting
+        # down the thread
+        self.connect_attempt_limit = max(1, connect_attempt_limit)
+        self._connection_tries = 0  # count of connection events; reset on success
+
+        # invalid until set in _on_queue_verified
+        self._connected_at: float | None = None
+
+        # list of times that channel was last closed; see _on_channel_closed
+        self._channel_closes: list[float] = []
+
+        # how long a time frame to keep previous channel close times
+        self.channel_close_window_s = channel_close_window_s
+
+        # how many times allowed to retry opening a channel in the above time
+        # window before giving up and shutting down the thread
+        self.channel_close_window_limit = channel_close_window_limit
+
+        self._mq_conn: pika.SelectConnection | None = None
+        self._thread_id = 0  # 0 == not yet started
+        self._mq_chan: Channel | None = None
+        self._published: dict[int, Future] = {}
+        self._outstanding: queue.Queue[tuple[Future, bytes]] = queue.Queue()
+        self._total_published = 0
+        self._delivery_tag_index = 0
+        self._stop_event = threading.Event()
+
+        # start closed ("connected" after connect)
+        self.status = RabbitPublisherStatus.closed
+
+        publish_kw = dict(**self.queue_info["queue_publish_kwargs"])
+        if "properties" in publish_kw:
+            publish_kw["properties"] = BasicProperties(**publish_kw["properties"])
+        self._publish_kwargs = publish_kw
+
+        super().__init__()
+
+    def __repr__(self):
+        return "{}<{}; o:{:,}; t:{:,}>".format(
+            type(self).__name__,
+            "✓" if self.status == RabbitPublisherStatus.connected else "✗",
+            self._outstanding.qsize(),
+            self._total_published,
+        )
+
+    def run(self) -> None:
+        log.debug("%r thread begins", self)
+        self._thread_id = threading.get_ident()
+
+        while (
+            not self._stop_event.is_set()
+            and self._connection_tries < self.connect_attempt_limit
+        ):
+            if self._mq_conn or self._connection_tries:
+                wait_for = random.uniform(0.5, 10)
+                msg = f"%r reconnecting in {wait_for:.1f}s."
+                log.debug(msg, self)
+                if self._connection_tries == self.connect_attempt_limit - 1:
+                    log.warning(f"{msg}  (final attempt)", self)
+                time.sleep(wait_for)
```

Comment:
Would be nice to replace the time.sleep with a wait on the `self._stop_event`

Reply:
Good call, thank you!